File: R/decompose.R
# 1: Set up
install.packages(c("tidyquant", "forecast", "ggplot2"))

# 1.1: Load required libraries
library(tidyquant)
library(forecast)
library(ggplot2)

# 2: Collect past five years of price data for AAPL and save it to an object named 'AAPL_df'
AAPL_df <- tq_get("AAPL", from = Sys.Date() - years(5), to = Sys.Date())

# 2.1: Plot and visualize AAPL price data
AAPL_price <- ggplot(AAPL_df, aes(x = date, y = close)) +
        geom_line(col = "blue", linewidth = 1.1) +
        labs(x = "Date", y = "Price", title = "AAPL Price") +
        theme_minimal()

# 2.2: View the plot
AAPL_price

# 3: Get the adjusted close prices and convert to time series object
AAPL_adj_close <- AAPL_df$adjusted
AAPL_ts <- ts(AAPL_adj_close, frequency = 252)

# 3.1: Perform time series decomposition using an additive model
ts_decomp_add <- decompose(AAPL_ts, type = "additive")

# 3.2: Plot the decomposition components
autoplot(ts_decomp_add) + ggtitle("Additive Model Decomposition")

# 4: Perform time series decomposition using LOESS (STL)
ts_decomp_loess <- stl(AAPL_ts, s.window = "periodic")

# 4.1: Plot the decomposition components
autoplot(ts_decomp_loess) + ggtitle("LOESS Decomposition")

# 5: Create a combined plot for decomposition
combined_plot <- cowplot::plot_grid(
        autoplot(ts_decomp_add) + ggtitle("Additive Model Decomposition"),
        autoplot(ts_decomp_loess) + ggtitle("LOESS Decomposition"),
        ncol = 2
)

# 5.1: Display the combined plot
print(combined_plot)


### SHINY APP: =================================================================================================
# Install Shiny if not already installed
install.packages("shiny")

# Load libraries
library(shiny)
library(tidyquant)
library(ggplot2)

# Define the Shiny UI
ui <- fluidPage(
        titlePanel("AAPL Visualization"),
        sidebarLayout(
                sidebarPanel(
                        selectInput("top_plot", "Select Top Plot:", choices = c("AAPL Price" = "price", "Additive Model Decomposition" = "additive", "LOESS Decomposition" = "loess")),
                        selectInput("bottom_plot", "Select Bottom Plot:", choices = c("Additive Model Decomposition" = "additive", "LOESS Decomposition" = "loess"))
                ),
                mainPanel(
                        plotOutput("top_plot_output"),
                        plotOutput("bottom_plot_output")
                )
        )
)

# Define the Shiny server
server <- function(input, output) {

        AAPL_df <- reactive({
                df <- tq_get("AAPL", from = Sys.Date() - years(5), to = Sys.Date())
                df
        })

        # Extract the time series decomposition components
        ts_decomp_add <- reactive({
                AAPL_adj_close <- AAPL_df()$adjusted
                ts_decomp_add <- decompose(ts(AAPL_adj_close, frequency = 252), type = "additive")
                ts_decomp_add
        })

        ts_decomp_loess <- reactive({
                AAPL_adj_close <- AAPL_df()$adjusted
                ts_decomp_loess <- stl(ts(AAPL_adj_close, frequency = 252), s.window = "periodic")
                ts_decomp_loess
        })

        output$top_plot_output <- renderPlot({})
        output$bottom_plot_output <- renderPlot({})

        observeEvent(input$top_plot, {
                if (input$top_plot == "price") {
                        p <- ggplot(AAPL_df(), aes(x = date, y = close)) +
                                geom_line(col = "blue", linewidth = 1.1) +
                                labs(x = "Date", y = "Price", title = "AAPL Price") +
                                theme_minimal()
                } else if (input$top_plot == "additive") {
                        p <- autoplot(ts_decomp_add()) +
                                ggtitle("Additive Model Decomposition")
                } else {
                        p <- autoplot(ts_decomp_loess()) +
                                ggtitle("LOESS Decomposition")
                }
                output$top_plot_output <- renderPlot({ p })
        })

        observeEvent(input$bottom_plot, {
                if (input$bottom_plot == "additive") {
                        p <- autoplot(ts_decomp_add()) +
                                ggtitle("Additive Model Decomposition")
                } else {
                        p <- autoplot(ts_decomp_loess()) +
                                ggtitle("LOESS Decomposition")
                }
                output$bottom_plot_output <- renderPlot({ p })
        })

}

# Run the Shiny app
shinyApp(ui = ui, server = server)

File: python/decompose.py
# 1: Install needed packages
!pip install yfinance
!pip install pandas
!pip install matplotlib
!pip install statsmodels

# 1.1: Load required libraries
import yfinance as yf
import pandas as pd
import matplotlib.pyplot as plt
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.seasonal import STL

# 2: Collect past 5 years of price data for AAPL and save it to an object named 'AAPL_df'
AAPL_df = yf.download('AAPL', start='2017-03-31', end='2022-03-31')

# 2.1: Plot and visualize AAPL price data
AAPL_price = AAPL_df[['Adj Close']].plot(figsize=(12, 6), xlabel='Date', ylabel='Price', title='AAPL Price')

# 2.2: View the plot
plt.show()

# 3: Get the adjusted close prices and convert to pandas series
AAPL_adj_close = AAPL_df["Adj Close"]

# 3.1: Perform time series decomposition using an additive model
ts_decomp_add = seasonal_decompose(AAPL_adj_close, model="additive", period=252)

# 3.2: Plot the decomposition components
ts_decomp_add.plot()
plt.suptitle("Additive Model Decomposition", y=0.99, fontsize=14)
plt.tight_layout()
plt.show()

# 4:  Perform time series decomposition using STL/LOESS method

ts_decomp_loess = STL(AAPL_adj_close, period=252).fit()

# 4.1: Plot the decomposition components
ts_decomp_loess.plot()
plt.suptitle("LOESS Decomposition", y=0.99, fontsize=14)
plt.tight_layout()
plt.show()

# 5: Prepare a grid for combined plot
fig, axes = plt.subplots(3, 2, sharex=True, figsize=(10, 12))
plt.tight_layout()
fig.subplots_adjust(top=0.95)

# 5.1: Additive Model Decomposition components
ts_decomp_add.trend.plot(ax=axes[0, 0], title="Trend")
ts_decomp_add.seasonal.plot(ax=axes[1, 0], title="Seasonality")
ts_decomp_add.resid.plot(ax=axes[2, 0], title="Residuals")
axes[0, 0].set_title("Additive Model Decomposition", fontsize=14, fontweight="bold")

# 5.2: LOESS Decomposition components
ts_decomp_loess.trend.plot(ax=axes[0, 1], title="Trend")
ts_decomp_loess.seasonal.plot(ax=axes[1, 1], title="Seasonality")
ts_decomp_loess.resid.plot(ax=axes[2, 1], title="Residuals")
axes[0, 1].set_title("LOESS Decomposition", fontsize=14, fontweight="bold")

# 5.3: Show the combined plot
plt.show()
